File: SwarmAI_tauri/src-tauri/src/models/timeline.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// TimelineItem
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimelineItem {
 pub id: String,
 pub turn_id: Option<Uuid>,
 pub date: DateTime<Utc>,
 pub content: TimelineContent,
}

// Constructors for each TimelineContent variant

impl TimelineItem {
 pub fn user(text: String) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::User(UserMessage { text, attachments: Vec::new(), hydra_heads: None }),
 }
 }

 pub fn assistant(text: &str) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Assistant(AssistantMessage { text: text.to_string(), is_streaming: false }),
 }
 }

 pub fn reasoning(text: &str) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Reasoning(ReasoningBlock { text: text.to_string(), is_streaming: false }),
 }
 }

 pub fn tool(call: ToolCall) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Tool(call),
 }
 }

 pub fn notice(message: String, level: NoticeLevel) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Notice(Notice { level, message }),
 }
 }

 pub fn turn_end(summary: TurnSummary) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: Some(summary.turn_id),
 date: Utc::now(),
 content: TimelineContent::TurnEnd(summary),
 }
 }

 pub fn approval(request: String) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Notice(Notice { level: NoticeLevel::Info, message: format!("Approval required: {}", request) }),
 }
 }

 pub fn plan(markdown: String) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Plan(ProposedPlan { markdown, state: PlanState::Drafting }),
 }
 }

 pub fn todos(steps: Vec<TodoStep>) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 turn_id: None,
 date: Utc::now(),
 content: TimelineContent::Todos(steps),
 }
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum TimelineContent {
 User(UserMessage),
 Assistant(AssistantMessage),
 Reasoning(ReasoningBlock),
 Tool(ToolCall),
 Plan(ProposedPlan),
 Todos(Vec<TodoStep>),
 Notice(Notice),
 TurnEnd(TurnSummary),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserMessage {
 pub text: String,
 pub attachments: Vec<Attachment>,
 pub hydra_heads: Option<Vec<u32>>,
}

impl UserMessage {
 pub fn is_from_hydra(&self) -> bool {
 self.hydra_heads.is_some()
 }
 pub fn is_hydra_report(&self) -> bool {
 !(self.hydra_heads.clone().unwrap_or_default().is_empty())
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FollowUpPrompt {
 pub id: Uuid,
 pub text: String,
 pub attachments: Vec<Attachment>,
 pub created_at: DateTime<Utc>,
}

impl FollowUpPrompt {
 pub fn is_empty(&self) -> bool {
 self.text.trim().is_empty() && self.attachments.is_empty()
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Attachment {
 pub id: Uuid,
 pub name: String,
 pub path: String,
 pub mime_type: String,
}

impl Attachment {
 pub fn is_image(&self) -> bool {
 self.mime_type.starts_with("image/")
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssistantMessage {
 pub text: String,
 pub is_streaming: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReasoningBlock {
 pub text: String,
 pub is_streaming: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToolKind {
 Command,
 Read,
 Edit,
 Search,
 Web,
 Mcp,
 Agent,
 Other,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ToolStatus {
 Running,
 Completed,
 Failed,
 Declined,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
 pub kind: ToolKind,
 pub title: String,
 pub detail: Option<String>,
 pub output: String,
 pub status: ToolStatus,
 pub exit_code: Option<i32>,
 pub edits: Vec<FileEdit>,
 pub started_at: DateTime<Utc>,
 pub finished_at: Option<DateTime<Utc>>,
}

impl ToolCall {
 pub const OUTPUT_LIMIT: usize = 60_000;

 pub fn finish(&mut self, status: ToolStatus) {
 self.status = status;
 if self.finished_at.is_none() {
 self.finished_at = Some(Utc::now());
 }
 }

 pub fn append_output(&mut self, text: &str) {
 self.output.push_str(text);
 self.trim_output();
 }

 pub fn set_output(&mut self, text: String) {
 self.output = text;
 self.trim_output();
 }

 fn trim_output(&mut self) {
 if self.output.len() > Self::OUTPUT_LIMIT {
 let keep = Self::OUTPUT_LIMIT / 2;
 let cut = self.output.len() - keep;
 self.output = format!("…{}", &self.output[cut..]);
 }
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileEdit {
 pub path: String,
 pub diff: Option<String>,
 pub additions: i32,
 pub deletions: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum PlanState {
 Drafting,
 Proposed,
 Accepted,
 Dismissed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProposedPlan {
 pub markdown: String,
 pub state: PlanState,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TodoStatus {
 Pending,
 Active,
 Done,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoStep {
 pub text: String,
 pub status: TodoStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NoticeLevel {
 Info,
 Warning,
 Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Notice {
 pub level: NoticeLevel,
 pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TurnStatus {
 Running,
 Completed,
 Interrupted,
 Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TurnSummary {
 pub turn_id: Uuid,
 pub status: TurnStatus,
 pub duration: f64,
 pub files_changed: u32,
 pub additions: u32,
 pub deletions: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TurnRecord {
 pub id: Uuid,
 pub index: u32,
 pub provider_turn_id: Option<String>,
 pub started_at: DateTime<Utc>,
 pub completed_at: Option<DateTime<Utc>>,
 pub status: TurnStatus,
 pub base_checkpoint: Option<String>,
 pub end_checkpoint: Option<String>,
 pub user_item_id: Option<String>,
 pub provider_diff: Option<String>,
 pub provider_anchor: Option<String>,
 pub touched_paths: Option<Vec<String>>,
 pub hydra_merged: bool,
}

impl TurnRecord {
 pub fn new(index: u32) -> Self {
 Self {
 id: Uuid::new_v4(),
 index,
 provider_turn_id: None,
 started_at: Utc::now(),
 completed_at: None,
 status: TurnStatus::Running,
 base_checkpoint: None,
 end_checkpoint: None,
 user_item_id: None,
 provider_diff: None,
 provider_anchor: None,
 touched_paths: None,
 hydra_merged: false,
 }
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextUsage {
 pub used_tokens: u64,
 pub window_tokens: Option<u64>,
}

impl ContextUsage {
 pub fn new(used: u64, window: u64) -> Self {
 Self {
 used_tokens: used,
 window_tokens: Some(window),
 }
 }

 pub fn fraction(&self) -> Option<f64> {
 match self.window_tokens {
 Some(w) if w > 0 => Some((self.used_tokens as f64 / w as f64).min(1.0)),
 _ => None,
 }
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadDocument {
 pub thread_id: Uuid,
 pub items: Vec<TimelineItem>,
 pub turns: Vec<TurnRecord>,
 pub usage: Option<ContextUsage>,
 pub follow_ups: Vec<FollowUpPrompt>,
}

impl ThreadDocument {
 pub fn new(thread_id: Uuid) -> Self {
 Self {
 thread_id,
 items: Vec::new(),
 turns: Vec::new(),
 usage: None,
 follow_ups: Vec::new(),
 }
 }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscriptMessage {
 pub role: String,
 pub content: String,
 pub timestamp: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadSummary {
 pub id: Uuid,
 pub name: String,
 pub project_id: Option<Uuid>,
 pub provider: crate::models::provider::ProviderKind,
 pub model: Option<String>,
 pub status: crate::models::ThreadStatus,
 pub created_at: DateTime<Utc>,
 pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ThreadDocumentSection {
 pub title: String,
 pub markdown: String,
 pub items: Vec<TimelineItem>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ThreadStatus {
 Idle,
 Running,
 Paused,
 Completed,
 Failed,
 Cancelled,
}

impl Default for ThreadStatus {
 fn default() -> Self {
 Self::Idle
 }
}

// ---------------------------------------------------------------------------
// TimelineEntry
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimelineEntry {
 pub id: String,
 pub item: TimelineItem,
 pub parent_entry: Option<String>,
}

impl TimelineEntry {
 pub fn new(item: TimelineItem) -> Self {
 Self {
 id: Uuid::new_v4().to_string(),
 item,
 parent_entry: None,
 }
 }
}

File: DroppyCode/Core/Models/Timeline.swift
import Foundation

struct TimelineItem: Codable, Identifiable, Hashable, Sendable {
    var id: String
    var turnID: UUID?
    var date: Date
    var content: Content

    init(id: String = UUID().uuidString, turnID: UUID?, date: Date = .now, content: Content) {
        self.id = id
        self.turnID = turnID
        self.date = date
        self.content = content
    }

    enum Content: Codable, Hashable, Sendable {
        case user(UserMessage)
        case assistant(AssistantMessage)
        case reasoning(ReasoningBlock)
        case tool(ToolCall)
        case plan(ProposedPlan)
        case todos([TodoStep])
        case notice(Notice)
        case turnEnd(TurnSummary)
    }

    /// A thinking block with nothing to read, as a redacted stream leaves behind.
    var isEmptyReasoning: Bool {
        guard case .reasoning(let block) = content else { return false }
        return block.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

extension TimelineItem {
    /// Model-written prose with its long dashes flattened, for threads stored by a
    /// build that cleaned text only as it arrived. The user's own words, tool output
    /// and notices are not the model's writing and are left exactly as they were.
    var cleanedOfEmDashes: TimelineItem {
        var item = self
        switch item.content {
        case .assistant(var message):
            message.text = TextCleanup.withoutEmDashes(message.text)
            item.content = .assistant(message)
        case .reasoning(var block):
            block.text = TextCleanup.withoutEmDashes(block.text)
            item.content = .reasoning(block)
        case .plan(var plan):
            plan.markdown = TextCleanup.withoutEmDashes(plan.markdown)
            item.content = .plan(plan)
        default:
            break
        }
        return item
    }
}

struct UserMessage: Codable, Hashable, Sendable {
    var text: String
    var attachments: [Attachment] = []
    /// Set when the message is Hydra's rather than the user's own words: the roster
    /// places of the heads reporting back, or empty for a note from Hydra itself.
    var hydraHeads: [Int]?
    /// Set when the message is the lead's brief to a head, shown as a pill rather
    /// than a plain bubble. Optional, so threads stored before it still decode.
    var hydraBrief: Bool?

    var isFromHydra: Bool { hydraHeads != nil }
    var isHydraBrief: Bool { hydraBrief == true }
}

/// One queued follow-up prompt. Sent as a direct user chat message once the
/// running turn finishes, so the user can steer the chat while it works and
/// stack several follow-ups that run one after another.
struct FollowUpPrompt: Codable, Hashable, Identifiable, Sendable {
    var id = UUID()
    var text: String
    var attachments: [Attachment] = []
    var createdAt = Date.now
    /// Prompts sharing a bundle sit together, show one number and go out as one
    /// message (see `ThreadRuntime.bundleFollowUp`).
    var bundleID: UUID? = nil

    var isEmpty: Bool {
        attachments.isEmpty && !text.contains { !$0.isWhitespace }
    }
}

struct Attachment: Codable, Hashable, Identifiable, Sendable {
    var id = UUID()
    var name: String
    var path: String
    var mimeType: String

    var url: URL { URL(fileURLWithPath: path) }
    var isImage: Bool { mimeType.hasPrefix("image/") }

    /// The images among the attachments with their bytes as base64, read and encoded off
    /// the main actor: a photo is megabytes, and its string a third larger again.
    @concurrent
    static func base64Images(_ attachments: [Attachment]) async -> [(image: Attachment, base64: String)] {
        attachments.compactMap { attachment in
            guard attachment.isImage, let data = try? Data(contentsOf: attachment.url) else { return nil }
            return (attachment, data.base64EncodedString())
        }
    }
}

struct AssistantMessage: Codable, Hashable, Sendable {
    var text: String
    var isStreaming = false
}

struct ReasoningBlock: Codable, Hashable, Sendable {
    var text: String
    var isStreaming = false
}

struct ToolCall: Codable, Hashable, Sendable {
    enum Kind: String, Codable, Sendable {
        case command
        case read
        case edit
        case search
        case web
        case mcp
        case agent
        case other
    }

    enum Status: String, Codable, Sendable {
        case running
        case completed
        case failed
        case declined
    }

    static let outputLimit = 60_000

    var kind: Kind
    var title: String
    var detail: String?
    var output = ""
    var status: Status = .running
    var exitCode: Int?
    var edits: [FileEdit] = []
    var startedAt = Date.now
    var finishedAt: Date?

    mutating func appendOutput(_ text: String) {
        output += text
        trimOutput()
    }

    mutating func setOutput(_ text: String) {
        output = text
        trimOutput()
    }

    mutating func finish(_ status: Status) {
        self.status = status
        if finishedAt == nil { finishedAt = .now }
    }

    private mutating func trimOutput() {
        guard output.utf8.count > Self.outputLimit else { return }
        output = "…" + String(output.suffix(Self.outputLimit / 2))
    }
}

struct FileEdit: Codable, Hashable, Sendable {
    var path: String
    var diff: String?
    var additions = 0
    var deletions = 0
}

struct ProposedPlan: Codable, Hashable, Sendable {
    enum State: String, Codable, Sendable {
        case drafting
        case proposed
        case accepted
        case dismissed
    }

    var markdown: String
    var state: State
}

struct TodoStep: Codable, Hashable, Sendable {
    enum Status: String, Codable, Sendable {
        case pending
        case active
        case done
    }

    var text: String
    var status: Status
}

struct Notice: Codable, Hashable, Sendable {
    enum Level: String, Codable, Sendable {
        case info
        case warning
        case error
    }

    var level: Level
    var message: String
    /// Set on the usage-limit notice alone: the chat whose auto-continue it belongs to.
    /// Its row draws the badge with the two controls while that chat waits (see `AutoContinue`).
    var limit: LimitNotice? = nil
}

/// The chat a usage-limit notice belongs to. `AutoContinue` marks its notice with this,
/// and the notice's row reads that chat's wait and outcome from it.
struct LimitNotice: Codable, Hashable, Sendable {
    var threadID: UUID
}

enum TurnStatus: String, Codable, Sendable {
    case running
    case completed
    case interrupted
    case failed
}

struct TurnSummary: Codable, Hashable, Sendable {
    var turnID: UUID
    var status: TurnStatus
    var duration: TimeInterval
    var filesChanged: Int
    var additions: Int
    var deletions: Int
    var changes: FileChangeSummary?
}

/// One outcome of the Hydra auto-merge for this thread, kept so the lead can be told the true merge state at the front of its next message.
struct HydraMergeRecord: Codable, Hashable, Sendable {
    enum Outcome: String, Codable, Sendable { case merged, failed, stray }
    var at: Date
    var outcome: Outcome
    var project: String?
    var label: String?
    var url: URL?
    var files: Int
    var detail: String?
}

struct TurnRecord: Codable, Identifiable, Hashable, Sendable {
    var id = UUID()
    var index: Int
    var providerTurnID: String?
    var startedAt = Date.now
    var completedAt: Date?
    var status: TurnStatus = .running
    var baseCheckpoint: String?
    var endCheckpoint: String?
    var userItemID: String?
    /// The provider's own diff, used when the workspace is not a git repository.
    var providerDiff: String?
    /// The provider's last message id in this turn, used to rewind a conversation.
    var providerAnchor: String?
    /// Repository-relative paths this turn's agent edited, so changes made elsewhere in the
    /// repository during the turn are never attributed to the thread. Nil on turns recorded before.
    var touchedPaths: [String]?
    /// This turn's work has landed on the branch: the auto-merge took it, and the next one
    /// starts from what came after it.
    var hydraMerged = false
}

extension TurnRecord {
    /// Read field by field, so a turn stored by an earlier build, which knew nothing of
    /// the newest of them, still reads as the turn it was rather than being dropped.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = container.value(.id, default: UUID())
        index = container.value(.index, default: 0)
        providerTurnID = container.value(.providerTurnID, default: nil)
        startedAt = container.value(.startedAt, default: Date.now)
        completedAt = container.value(.completedAt, default: nil)
        status = container.value(.status, default: .running)
        baseCheckpoint = container.value(.baseCheckpoint, default: nil)
        endCheckpoint = container.value(.endCheckpoint, default: nil)
        userItemID = container.value(.userItemID, default: nil)
        providerDiff = container.value(.providerDiff, default: nil)
        providerAnchor = container.value(.providerAnchor, default: nil)
        touchedPaths = container.value(.touchedPaths, default: nil)
        hydraMerged = container.value(.hydraMerged, default: false)
    }
}

struct ContextUsage: Codable, Hashable, Sendable {
    var usedTokens: Int
    var windowTokens: Int?

    var fraction: Double? {
        guard let windowTokens, windowTokens > 0 else { return nil }
        return min(1, Double(usedTokens) / Double(windowTokens))
    }
}

struct ThreadDocument: Codable, Equatable, Sendable {
    var threadID: UUID
    var items: [TimelineItem] = []
    var turns: [TurnRecord] = []
    var usage: ContextUsage?
    var followUps: [FollowUpPrompt] = []
    var hydraMerges: [HydraMergeRecord] = []
    /// The forwarded chat this thread continues, when the user opened it with one. Nil for
    /// a thread that started here. Optional, so threads stored before it still decode.
    var continuation: ThreadContinuation? = nil

    init(threadID: UUID) {
        self.threadID = threadID
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        threadID = try container.decode(UUID.self, forKey: .threadID)
        items = container.value(.items, default: [Lenient<TimelineItem>]()).compactMap(\.value)
        turns = container.value(.turns, default: [Lenient<TurnRecord>]()).compactMap(\.value)
        usage = container.value(.usage, default: nil)
        followUps = container.value(.followUps, default: [Lenient<FollowUpPrompt>]()).compactMap(\.value)
        hydraMerges = container.value(.hydraMerges, default: [Lenient<HydraMergeRecord>]()).compactMap(\.value)
        continuation = container.value(.continuation, default: nil)
    }
}
